woocommerce/woocommerce
# POC: Product template switching within the blocks editor

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

* Adds the classes for containing and implementing template types
* Adds hooks to allow switching between templates
* Adds a component to handle layout switching for templates

This PR does not yet address:

* Fetching templates async
* Reusing template pieces within templates
* Logic for determining when to switch between templates

>Fetching templates async

This is the most relevant to this PR and something that we need to discuss.  We can fetch these templates async or preload them all on page load.

1. Without preloading these, the UX might be clunky and slow when switching between templates.  We would need to show some kind of loading indicator while the other templates are fetched.  I'm not sure if this is a dealbreaker cc @jarekmorawski 
2. If we preload, we'll need to do this on every WCA page which is quite a bit of data to load for pages that may not even need this
3. The hybrid solution (which I'd lean towards) is to fetch all editor settings from an endpoint when switching to a product page.  This may add a very minor delay to the initial load of products, but only needs to occur once and then template switching should be seamless.

Closes #38277  .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Visit WooCommerce->Settings->Advanced->Features and turn on the new product block editor.
4. Checkout the branch `add/38277-test`
5. Navigate to the Products -> Add new screen
6. Note that the template shows correctly
7. Click the "Toggle template" button
8. Check that the template is switched (only the "General" tab will be shown)
